ledger/ledger-api-integration-tests/src/test/lib/scala/com/digitalasset/platform/apitesting/CommandTransactionChecks.scala

```diff
@@ -1161,9 +944,254 @@ class CommandTransactionChecks(
+      ctx: LedgerContext,
+      command: SubmitRequest,
+      transactionFilter: TransactionFilter,
+      filterCid: Boolean = true): Future[Seq[Transaction]] = {
+    submitAndListenForTransactionResultsOfCommand(ctx, command, transactionFilter, filterCid)
+  }
+
+  private def submitAndListenForTransactionResultsOfCommand(
+      ctx: LedgerContext,
+      submitRequest: SubmitRequest,
+      transactionFilter: TransactionFilter,
+      filterCid: Boolean = true): Future[immutable.Seq[Transaction]] = {
+    val commandId = submitRequest.getCommands.commandId
+    for {
+      txEndOffset <- ctx.testingHelpers.submitSuccessfullyAndReturnOffset(submitRequest)
+      transactions <- listenForTransactionResultOfCommand(
+        ctx,
+        transactionFilter,
+        if (filterCid) Some(commandId) else None,
+        txEndOffset)
+    } yield {
+      transactions
+    }
+  }
+
+  private def listenForTransactionResultOfCommand(
+      ctx: LedgerContext,
+      transactionFilter: TransactionFilter,
+      commandId: Option[String],
+      txEndOffset: LedgerOffset): Future[immutable.Seq[Transaction]] = {
+    ctx.transactionClient
+        .getTransactions(
+          txEndOffset,
+          None,
+          transactionFilter
+        )
+        .filter(x => commandId.fold(true)(cid => x.commandId == cid))
+        .take(1)
+        .takeWithin(3.seconds)
+        .runWith(Sink.seq)
+  }
+
+  private def simpleExerciseWithListenerForTransactions(
+      ctx: LedgerContext,
+      commandId: String,
+      submitter: String,
+      listener: String,
+      template: Identifier,
+      contractId: String,
+      choice: String,
+      arg: Value
+  ): Future[Seq[Transaction]] = {
+    submitAndListenForTransactionResultOfCommand(
+      ctx,
+      submitRequestWithId(ctx, cid(commandId))
+          .update(
+            _.commands.commands :=
+                List(ExerciseCommand(Some(template), contractId, choice, Some(arg)).wrap),
+            _.commands.party := submitter
+          ),
+      TransactionFilter(Map(listener -> Filters.defaultInstance)),
+      false
+    )
+  }
+
+  private def transactionsFromSimpleExercise(
+      ctx: LedgerContext,
+      commandId: String,
+      submitter: String,
+      template: Identifier,
+      contractId: String,
+      choice: String,
+      arg: Value): Future[Seq[Transaction]] =
+    simpleExerciseWithListenerForTransactions(
+      ctx,
+      commandId,
+      submitter,
+      submitter,
+      template,
+      contractId,
+      choice,
+      arg)
+
+  private def assertCommandFailsWithCode(
+      ctx: LedgerContext,
+      submitRequest: SubmitRequest,
+      expectedErrorCode: Code,
+      expectedMessageSubString: String): Future[Assertion] = {
+    for {
+      ledgerEnd <- ctx.transactionClient.getLedgerEnd
+      completion <- submitCommand(ctx, submitRequest)
+      // TODO(FM) in the contract keys test this hangs forever after expecting a failedExercise.
```

Comment:
@gaboraranyossy-da do you have an idea on why this might happe?

Reply:
I need more context on this, let's chat when you have time.